app/src/test/java/chat/rocket/android/authentication/loginoptions/presentation/LoginOptionsPresenterTest.kt

```diff
@@ -52,4 +72,57 @@ class LoginOptionsPresenterTest {
         loginOptionsPresenter.toLoginWithEmail()
         verify(navigator).toLogin(currentServer)
     }
+
+    @Test
+    fun `check account is saved`() {
+        loginOptionsPresenter.authenticateWithOauth(OAUTH_TOKEN, OAUTH_SECRET)
+        val method = loginOptionsPresenter.javaClass.getDeclaredMethod("saveAccount", String::class.java)
+        method.isAccessible = true
+        val parameters = arrayOfNulls<Any>(1)
+        parameters[0] = USERNAME
+        method.invoke(loginOptionsPresenter, *parameters)
+        verify(saveAccountInteractor).save(account)
+    }
+
+    @Test
+    fun `setup connection info`() {
+        val method = loginOptionsPresenter.javaClass.getDeclaredMethod("setupConnectionInfo", String::class.java)
+        method.isAccessible = true
+        val parameters = arrayOfNulls<Any>(1)
+        parameters[0] = currentServer
+        method.invoke(loginOptionsPresenter, *parameters)
+        assertEquals(parameters[0], currentServer)
+    }
+
+    @Test
+    fun `authenticate user with Oauth`() {
+        kotlinx.coroutines.runBlocking {
+            val result = loginOptionsPresenter.authenticateWithOauth(OAUTH_TOKEN, OAUTH_SECRET)
+            assertEquals(result, Unit)
+        }
+    }
+
+    @Test
+    fun `authenticate user with Cas`() {
+        kotlinx.coroutines.runBlocking {
+            val result = loginOptionsPresenter.authenticateWithCas(CAS_TOKEN)
+            assertEquals(result, Unit)
+        }
+    }
+
+    @Test
+    fun `authenticate user with Saml`() {
+        kotlinx.coroutines.runBlocking {
+            val result = loginOptionsPresenter.authenticateWithSaml(SAML_TOKEN)
+            assertEquals(result, Unit)
+        }
+    }
+
+    @Test
+    fun `authenticate user with Deeplink`() {
```

Comment:
I'm seeing an exception being thrown for all of these coroutine tests, and yet, after the exception the test passes.... do you know what's going on?

```chat.rocket.android.authentication.loginoptions.presentation.LoginOptionsPresenterTest > authenticate user with Deeplink STANDARD_ERROR
    Exception in thread "DefaultDispatcher-worker-7 @coroutine#16" kotlin.UninitializedPropertyAccessException: lateinit property client has not been initialized
        at chat.rocket.android.authentication.loginoptions.presentation.LoginOptionsPresenter.access$getClient$p(LoginOptionsPresenter.kt:44)
        at chat.rocket.android.authentication.loginoptions.presentation.LoginOptionsPresenter$doAuthentication$1$token$1.invokeSuspend(LoginOptionsPresenter.kt:112)
        at chat.rocket.android.authentication.loginoptions.presentation.LoginOptionsPresenter$doAuthentication$1$token$1.invoke(LoginOptionsPresenter.kt)
        at chat.rocket.android.util.IOKt.retryIO(IO.kt:25)
        at chat.rocket.android.util.IOKt.retryIO$default(IO.kt:18)
        at chat.rocket.android.authentication.loginoptions.presentation.LoginOptionsPresenter$doAuthentication$1.invokeSuspend(LoginOptionsPresenter.kt:110)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:233)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)

chat.rocket.android.authentication.loginoptions.presentation.LoginOptionsPresenterTest > authenticate user with Deeplink PASSED```

Reply:
We still do not have a solution to unit test coroutine methods, Shailesh's solution has solved the null context issue but it is not helpful in testing coroutines methods. I had also discussed with him personally. At the same time, these coroutines methods are very well tested with UI tests so there is no problem. Also, we are now testing the private methods individually using Java reflection which had reduced our dependence on coroutines methods. I have removed the unit tests for coroutines methods as they are not working as expected.